README.md

```diff
@@ -264,5 +265,75 @@ result = downloads.join(
 ```
 The window function version is usually easier to get right and is usually more concise.
 
+
+### If feasible, `select` columns explicitly and avoid chaining `withColumn`
+```python
+previous_ips = F.coalesce("previous_ips", F.array([]))
+
+# Option A: good
+df.select(
+    "ip_addresses",
+    previous_ips.alias("previous_ips"),
+    F.array_except("ip_addresses", previous_ips).alias("new_ips"),
+    F.col("user_id").alias("user_name")
+)
+
+# Option B: bad - F.array_except uses the original "previous_ips" column
+# In this case, some values in the "new_ips" column may be null.
+df.select(
+    "ip_addresses",
+    previous_ips.alias("previous_ips"),
+    F.array_except("ip_addresses", "previous_ips").alias("new_ips"),
+    F.col("user_id").alias("user_name")
+)
+
+# Option C: bad - not explicit, selects two different columns both named "previous_ips"
+df.select(
+    "*",
+    previous_ips.alias("previous_ips"),
+    F.array_except("ip_addresses", previous_ips).alias("new_ips"),
+    F.col("user_id").alias("user_name"),
+)
+
+# Option D: bad - involves an additional Project stage
+df.withColumn(
+    "previous_ips", previous_ips
+).withColumn(
+    "new_ips", F.array_except("ip_addresses", "previous_ips")
+).withColumnRenamed(
+    "user_id", "user_name"
+)
+```
+In addition to improving readability, using `select` to create new columns provides an additional advantage: `select` ensures that all of the columns are created in only one `Project` stage, whereas `withColumn` creates additional internal projection stages (see the caveat in the [PySpark documentation](https://spark.apache.org/docs/latest/api/python/reference/api/pyspark.sql.DataFrame.withColumn.html)). Using a single `select` statement simplifies the physical plan and will likely improve query performance. 
```

Comment:
Maybe a slight rewording to indicate that this is an example physical plan that Spark built in your test? In general, the same code can lead to different physical plans in different situations.

Reply:
Good idea - I've updated this description to clarify that point.